build: typescript 3.7 support

andrius-pra opened the pull request:
This PR updates TypeScript version to 3.7 while retaining compatibility with TS3.6.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #33596


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

andrius-pra committed:
build: typescript 3.7 support

This PR updates TypeScript version to 3.7 while retaining compatibility with TS3.6.

alfaproject commented on packages/zone.js/test/common/fetch.spec.ts:
```diff
@@ -150,7 +150,7 @@ describe(
         it('cancel fetch should invoke onCancelTask',
            ifEnvSupportsWithDone('AbortController', (done: DoneFn) => {
-             if (isSafari) {
+             if (isSafari()) {
```
<3

andrius-pra commented on packages/compiler-cli/ngcc/src/packages/configuration.ts:
```diff
@@ -266,5 +266,6 @@ function findSatisfactoryVersion(
     return configs[0];
   }
-  return configs.find(config => satisfies(version, config.versionRange)) || null;
+  const options = {includePrerelease: true};
```
fix for this [test](https://github.com/angular/angular/blob/master/packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts#L659-L692). 
`satisfies('9.0.0-rc.1+34.sha-0e36204.with-local-changes', '*')` returns false without includePrerelease option.

andrius-pra commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1277,7 +1277,8 @@ function saveNameToExportMap(
     }
-    if ((def as ComponentDef<any>).template) exportsMap[''] = index;
+
+    if (isComponentDef(def)) exportsMap[''] = index;
```
Fix for:
```
error TS2774: This condition will always return true since the function is always defined. Did 
you mean to call it instead?
```

andrius-pra commented on packages/platform-server/src/tokens.ts:
```diff
@@ -32,4 +32,4 @@ export const INITIAL_CONFIG = new InjectionToken<PlatformConfig>('Server.INITIAL
  */
 export const BEFORE_APP_SERIALIZED =
-    new InjectionToken<Array<() => void>>('Server.RENDER_MODULE_HOOK');
+    new InjectionToken<Array<() => void | Promise<void>>>('Server.RENDER_MODULE_HOOK');
```
Fixes error in this line [line](https://github.com/angular/angular/blob/1d429b216556911edc4b0675ece4cb9081967155/packages/platform-server/src/utils.ts#L57)
```
asyncPromises.push(callbackResult);
                   ~~~~~~~~~~~~~~
Argument of type 'void' is not assignable to parameter of type 'Promise<any>'.
```

andrius-pra commented:
@filipesilva, I have added you as co-author because of this PR contains changes from #32962.

andrius-pra commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -823,3 +821,12 @@ function getFromSymbolTable(symbolTable: ts.SymbolTable, key: string): ts.Symbol
   return symbol;
 }
+
+function getTypeArguments(
```
The `typeArguments` property has been removed from the `TypeReference` interface in Typescipt 3.7. we should use the `getTypeArguments` function on `TypeChecker` instances. See [#32962](https://github.com/angular/angular/pull/32962#issuecomment-537524408) (comment)

gkalpak commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -673,26 +675,20 @@ function getBuiltinTypeFromTs(kind: BuiltinType, context: TypeContext): ts.Type
-        text: node.getText(),
-      };
-      setParents(<any>{kind: ts.SyntaxKind.ExpressionStatement, expression: numeric}, node);
+      const numeric = <ts.NumericLiteral>{kind: ts.SyntaxKind.NumericLiteral, text: '1'};
```
1. `text: '1'`? Shouldn't this be `text: node.getText()`? Or is it that the value doesn't make a difference, because we only care about the type?
2. Why not `ts.createNumericLiteral(node.getText())`?

(Same below.)

andrius-pra replied:
1. We only need a type and this type will cached [here](https://github.com/angular/angular/blob/master/packages/compiler-cli/src/diagnostics/typescript_symbols.ts#L96-L102). I can't use `text: node.getText()` because of this error:
```
Error: Debug Failure. False expression: This failure could trigger https://github.com/Microsoft/TypeScript/issues/20809
    ...
        at checkNumericLiteralValueSize (node_modules/typescript/lib/typescript.js:65912:56)
    ...
        at Object.getTypeAtLocation (node_modules/typescript/lib/typescript.js:33327:31)
        at getBuiltinTypeFromTs (packages/compiler-cli/src/diagnostics/typescript_symbols.ts:680:22)
    ...
        at UserContext.<anonymous> (packages/language-service/test/diagnostics_spec.ts:338:26)
```

andrius-pra replied:
2. `checker.getTypeAtLocation(typeNode)` will always returns `errorType` when parameter `typeNode` has `NodeFlags.Synthesized` flag. All `ts.create*` methods adds this flag. So we can use `ts.create*` method only for child nodes.

gkalpak reviewed (commented):
💯 :tada: ❤️

andrius-pra commented on aio/package.json:
```diff
@@ -165,7 +165,7 @@
     "ts-node": "^8.4.1",
     "tslint": "^5.20.0",
-    "typescript": "~3.6.4",
+    "typescript": "~3.7.2",
```
I reverted this change.

andrius-pra commented on packages/compiler-cli/src/diagnostics/typescript_symbols.ts:
```diff
@@ -673,26 +675,20 @@ function getBuiltinTypeFromTs(kind: BuiltinType, context: TypeContext): ts.Type
           <ts.Node><any>{
             kind: ts.SyntaxKind.VoidExpression,
-            expression: <ts.Node>{kind: ts.SyntaxKind.NumericLiteral}
+            expression: ts.createNumericLiteral('0')
```
Fix for : 
```
 Message:
    TypeError: Cannot read property 'length' of undefined
  Stack:
        at checkNumericLiteralValueSize (node_modules/typescript/lib/typescript.js:65901:77)
    ...
        at Object.getTypeAtLocation (node_modules/typescript/lib/typescript.js:33327:31)
        at getBuiltinTypeFromTs (packages/compiler-cli/src/diagnostics/typescript_symbols.ts:688:22)
    ...
        at UserContext.<anonymous> (packages/language-service/test/diagnostics_spec.ts:622:32)
```

gkalpak reviewed (commented):
BTW, I can see several more `.typeArguments` accesses through-out the codebase.
Shouldn't all these occurrences be converted to use `checker.getTypeArguments()` if available?

andrius-pra commented:
`.typeArguments` removed only from ts.TypeReference node (microsoft/TypeScript#33693). All these occurrences are converted to use `checker.getTypeArguments()`.

IgorMinar commented:
hi @andrius-pra, thanks for working on this PR - I just want to set the expectations right that we won't be merging this PR until we work on 9.1. I estimate that we will come back to this PR and be able to merge it in December, but unlikely sooner than that.

I'm grateful for your help, and we could use this PR in the future to get typescript 3.7 support into 9.1, but that will require cooperation with the CLI team and extensive testing that we no longer have time in 9.0 release cycle.

tony commented:
@IgorMinar I wanted to chime in and say I think that message you wrote to @andrius-pra was very kind and professional! :blush: In both how it mentioned release timings / priorities / logistics but also gratitude for the contribution and time put into Andrius work.  This is a big project with a lot of moving parts, and being graceful like that was a nice thing to see.

Look forward to seeing TypeScript 3.7 support in Angular!

Svetomechc commented:
@IgorMinar no 3.7 in 9.0 then?

filipesilva commented:
Heya @andrius-pra, just wanted to let you know that I'm looking at getting this PR green and ready to be merged. I'll do that by adding some commits to your branch directly, which I can already do because of the [allow edits from maintainers checkbox](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) on PR creation.

filipesilva committed:
test: add yarn.lock to typings_test_ts37

filipesilva committed:
build: update tsickle to 0.38.0

This is the tsickle version that supports TypeScript 3.7.

filipesilva committed:
test(docs-infra): compile only app for examples with no e2e

With TS 3.7, these examples were running into the error below (e.g. on https://circleci.com/gh/angular/angular/574906#tests/containers/0):

```
============== AIO example output for: /home/circleci/ng/aio/content/examples/observables/
running: yarn tsc --project ./
$ /home/circleci/ng/aio/content/examples/observables/node_modules/.bin/tsc --project ./
../../../tools/examples/shared/node_modules/protractor/built/index.d.ts(5,10): error TS2440: Import declaration conflicts with local declaration of 'PluginConfig'.
../../../tools/examples/shared/node_modules/protractor/built/index.d.ts(5,24): error TS2440: Import declaration conflicts with local declaration of 'ProtractorPlugin'.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
completed: yarn tsc --project ./
```

This happened because of https://github.com/angular/protractor/issues/5348.

It's unclear why this typings problem does not affect `ng e2e` runs, and only affects `tsc` runs.

For now it seems sensible to alter the tests to compile only the app and not the e2e, since the intent of https://github.com/angular/angular/commit/2cc954d5a54b2d9a0bee3679f6d4f9473389f32e was never to verify the correctness of the e2e in the first place.

We still need a release of protractor that supports TS 3.7 though, but at least it doesn't seem to block our update proper.

filipesilva committed:
build(docs-infra): use typescript 3.7

filipesilva commented:
@googlebot I consent.

filipesilva commented:
It's noteworthy that protractor itself doesn't seem to support TS 3.7 (https://github.com/angular/protractor/issues/5348) because of the ["Local and Imported Type Declarations Now Conflict" breaking change](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#local-and-imported-type-declarations-now-conflict), but that doesn't seem to affect anything but a few example projects that ran `tsc --project ./`. Not really sure why. 

I changed that command to compile the app and not the e2e since that was the original intention anyway. But we should still push for a Protractor release that supports TS 3.7.

filipesilva committed:
test: use ts 3.7 on material unit tests

filipesilva committed:
test: fix path to dist zone.js

filipesilva committed:
test: save material ts override as a devdep

filipesilva committed:
test: also install tsickle 0.38 on material

filipesilva commented on .circleci/config.yml:
```diff
@@ -764,6 +764,16 @@ jobs:
+          # While built Angular packages support both TS 3.6 and 3.7, @angular/bazel itself only
+          # supports TS 3.7 because TSickle 0.38.0 only supports TS 3.7.
+          # It shouldn't be necessary to install tsickle, but it's referenced in material and
+          # doesn't seem to be in the dependencies. Unsure how this worked before.
```
https://github.com/angular/components/pull/18116 will reference it in the dependencies. It was missing.

filipesilva commented:
Currently blocked on making the `ci/circleci: material-unit-tests` job green.

devversion committed:
ci: update material-unit-tests commit

Updates the commit SHA the `material-unit-tests` CircleCI
job runs against. We need to include a commit that makes
the node module installation more determinisitc (i.e. ensuring
that `tsickle` is always hoisted at the node module root).

https://github.com/angular/components/commit/31a50f7ad2947d9ceba99a64717ffa263a2a1d56

devversion committed:
ci: overwrite angular/components yarn typescript resolution

The components repository has a Yarn resolution to ensure that
dgeni-packages uses a specific TypeScript version. This resolution
causes the specified TS version to be considered as candidate for the
`@angular/bazel` peer dependency. Ultimately, Yarn decides to use
the TypeScript version from the resolution for `@angular/bazel`,
and builds will fail due to a version mismatch.

This is because `tsickle` will use the hoisted top-level TS
version (set to `3.7.4` ), while `@angular/bazel` uses the
version from the resolution (at the time of writing: v3.6.4)

devversion commented:
@googlebot I consent.

filipesilva commented on .circleci/config.yml:
```diff
@@ -766,14 +766,14 @@ jobs:
           # doesn't seem to be in the dependencies. Unsure how this worked before.
           # https://github.com/angular/components/blob/636ae5c8754008bfcf4b2f3953c6b2c5a7e27a1a/tools/bazel/BUILD.bazel#L12
-          command: yarn --ignore-engines --cwd ${MATERIAL_REPO_TMP_DIR} add typescript@3.7.4 tsickle@0.38.0 --dev
+          command: yarn --ignore-engines --cwd ${MATERIAL_REPO_TMP_DIR} add typescript@3.7.4 --dev
```
Why is Tsickle removed here? Tsickle 0.38 only works with TS 3.7 and should be updated.

devversion replied:
The comment above sounded like it was only needed because we didn't specify the dependency in `angular/components`. This has been fixed, so I removed it. Though, based on what you said, it sounds like we still need to manually update `tsickle`.

andrius-pra committed:
ci: update material-unit-tests commit

Updates the SHA that will be tested against in the `material-unit-tests` job to the latest commit in the components repository. SHA 71955d2e194bfc5561f25daea16e68af266d6ff9 is needed in order to compile repository with typescript 3.7

filipesilva commented on packages/language-service/src/typescript_symbols.ts:
```diff
@@ -313,8 +314,13 @@ class TypeWrapper implements Symbol {
+    let typeArguments: ReadonlyArray<ts.Type>|undefined;
+    if (this.context.checker.getTypeArguments)
+      typeArguments = this.context.checker.getTypeArguments(typeReference);
+    typeArguments = typeReference.typeArguments;
```
@andrius-pra should this be in a else clause?

filipesilva replied:
Added a commit with the else clause.

andrius-pra replied:
yes, it should be in else clause.

filipesilva committed:
fixup! build: typescript 3.7 support

gkalpak commented on aio/package.json:
```diff
@@ -165,7 +165,7 @@
     "ts-node": "^8.4.1",
     "tslint": "^5.20.0",
-    "typescript": "~3.6.4",
+    "typescript": "~3.7.4",
```
A little surprised this works (e.g. I would expect ngcc from the currently installed @angular/compiler-cli@9.0.0-rc.3 to fail), but if CI is happy, so am I :grin:

filipesilva replied:
Oh jeez, I was also expecting tests to fail and didn't look this up. It's really weird that the TS compilation doesn't show errors on these....

andrius-pra replied:
I reverted this change.

gkalpak reviewed (dismissed):
:tada:

Unless I am missing something, in order for everything to work correctly we need to account for the `typeArguments` breaking change in more places (e.g. [here][1] and [there][2] just to give a couple of examples - there are more places like this).
It is quite worrying that no tests are failing (unless, again, I am missing something) 😕 

[1]: https://github.com/angular/angular/blob/280aae6fc/packages/compiler-cli/ngcc/src/analysis/module_with_providers_analyzer.ts#L52
[2]: https://github.com/angular/angular/blob/280aae6fc/packages/compiler-cli/src/metadata/evaluator.ts#L469

filipesilva commented:
Oh geez, I also expected compilation to fail and didn't look these up actually.

Hmm I looked these up with TS 3.7 installed and they are of type `ts.NodeWithTypeArguments.typeArguments?`. Which sounds like it should have type arguments. I guess that's why, it's a different type.

andrius-pra commented:
The `typeArguments` property has been removed only from the [ts.TypeReference](https://github.com/microsoft/TypeScript/commit/250d5a8229e17342f36fe52545bb68140db96a2e#diff-2a68596a5960538c5c8733bc458de56eL2389-L2392) interface in typescript v3.7.0-beta. This type is only used in `typescript_symbols.ts` file.

IgorMinar commented:
I kicked off a presubmit for g3 on this PR: http://test/OCL:289739402:BASE:289739456:1579043225284:3be9296c

IgorMinar commented on integration/typings_test_ts37/package.json:
```diff
@@ -0,0 +1,30 @@
+    "@angular/upgrade": "file:../../dist/packages-dist/upgrade",
+    "@types/jasmine": "2.5.41",
+    "rxjs": "file:../../node_modules/rxjs",
+    "typescript": "3.7.3",
```
I'll push a fixup commit to this PR to fix this.

IgorMinar reviewed (approved):
lgtm! thanks @andrius-pra @filipesilva and @gkalpak !!!

I'll push the one fixup commit here and if the g3 presubmit is good we can land this.

IgorMinar committed:
test: update typings test to use ts 3.7.4 instead of 3.7.3

Just to be consistent.

IgorMinar commented:
merge-assistance: global approval

StefanKern commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -228,19 +229,19 @@ export declare class NgForOf<T, U extends NgIterable<T> = NgIterable<T>> impleme
     count: number;
-    readonly even: boolean;
-    readonly first: boolean;
+    get even(): boolean;
```
Why did you switch from `readonly` to `get`?
- curious angular developer

filipesilva replied:
We didn't, at least intentionally. 

These `.d.ts` files in `tools/public_api_guard/common` are snapshots of the types for the Angular API created by `ts-api-guardian`. The purpose of that tools is to tell us when the public TS api changed. This is useful to detect cases where we did some code change that accidentally changed the API, in which case we revert the change.

In this case however, TS started emitting different types so we had to update the snapshots. TS does not follow semver and can be expected to emit different type output between minors, so it is ok to update the snapshots.
